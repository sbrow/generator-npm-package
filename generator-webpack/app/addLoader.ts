import {
    createArrayLiteral,
    Identifier,
    isArrayLiteralExpression,
    Node,
    ObjectLiteralExpression,
    SourceFile,
    SyntaxKind,
    transform,
    TransformerFactory,
    visitEachChild,
    visitNode,
    Visitor,
} from "typescript";
import { RuleSetRule } from "webpack";

import {
    getByPath,
    getExports,
    getIdentifier,
    getProperty,
    toAST,
} from "./ast";

export function loaderToAST(loader: RuleSetRule): [Node, SourceFile] {
    const loaderAST = toAST(loader);
    const tree = [
        SyntaxKind.VariableStatement,
        SyntaxKind.VariableDeclarationList,
        SyntaxKind.VariableDeclaration,
        SyntaxKind.ObjectLiteralExpression,
    ];
    const node = getByPath(loaderAST, tree);
    return [node, loaderAST];
}

function SingleReplaceTransformer<T extends Node>(
    original: Node,
    replacement: Node,
): TransformerFactory<T> {
    return context => {
        const visit: Visitor = node => {
            if (node === original) {
                return replacement;
            }
            return visitEachChild(node, child => visit(child), context);
        };

        return node => visitNode(node, visit);
    };
}
export function addLoader(
    loader: RuleSetRule,
    sourceFile: SourceFile,
): SourceFile {
    const configVar = getConfig(sourceFile);
    const mod = getProperty(
        sourceFile,
        configVar,
        "module",
    ) as ObjectLiteralExpression;

    const rules = getProperty(sourceFile, mod, "rules");
    if (isArrayLiteralExpression(rules)) {
        const arr = createArrayLiteral(
            [
                ...rules.elements,
                { ...loaderToAST(loader)[0], _expressionBrand: true },
            ],
            true,
        );
        return updateNode(sourceFile, rules, arr);
    }
}
/**
 * @remarks
 * Works, but eliminates whitespace.
 */
function updateNode(
    sourceFile: SourceFile,
    target: Node,
    replacement: Node,
): SourceFile {
    const { transformed } = transform(sourceFile, [
        SingleReplaceTransformer(target, replacement),
    ]);
    return transformed[0];
}

function getConfig(
    sourceFile: SourceFile,
): ObjectLiteralExpression | undefined {
    const ID = getExports(sourceFile)[0] as Identifier;
    return (getIdentifier(
        sourceFile,
        ID,
    ) as unknown) as ObjectLiteralExpression;
}
