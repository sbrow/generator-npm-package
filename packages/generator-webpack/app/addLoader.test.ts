import {
    createPrinter,
    createSourceFile,
    EmitHint,
    Printer,
    ScriptTarget,
} from "typescript";
import { addLoader, loaderToAST } from "./addLoader";

let printer: Printer;

beforeEach(() => {
    printer = createPrinter({});
});

describe(addLoader.name, () => {
    const source = `const baseConfig = {
    module: {
        rules: [
            {foo: "bar"},
        ],
    },
};

module.exports = baseConfig;`;
    // tslint:disable-next-line: mocha-no-side-effect-code
    it("Updates rules", () => {
        const sourceFile = createSourceFile(
            "test.ts",
            source,
            ScriptTarget.Latest,
        );
        const got = addLoader(
            {
                use: "ts-loader",
            },
            sourceFile,
        );
        expect(
            printer.printNode(EmitHint.Unspecified, got, sourceFile),
        ).toMatchSnapshot();
    });
});
describe(loaderToAST.name, () => {
    // tslint:disable-next-line: mocha-no-side-effect-code
    it.each([
        {
            use: "ts-loader",
        },
        {
            exclude: [/node_modules/],
            use: "ts-loader",
        },
    ])("Works", (...args: Parameters<typeof loaderToAST>) => {
        const [object] = args;
        const [node, sourceFile] = loaderToAST(object);
        expect(
            printer.printNode(EmitHint.Unspecified, node, sourceFile),
        ).toMatchSnapshot();
    });
});
