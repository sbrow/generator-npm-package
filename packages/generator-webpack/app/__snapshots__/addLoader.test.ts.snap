// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addLoader Updates rules 1`] = `
"const baseConfig = {
    module: {
        rules: [
            { foo: \\"bar\\" },
            { use: \\"ts-loader\\" }
        ],
    },
};
module.exports = baseConfig;
"
`;

exports[`loaderToAST Works 1`] = `"{ use: \\"ts-loader\\" }"`;

exports[`loaderToAST Works 2`] = `"{ exclude: [/node_modules/], use: \\"ts-loader\\" }"`;
